### ISSUE TITLE
@ngrx/data - add entity op actions for offline support

### ISSUE THREAD

**Issue by AdditionAddict** [ngrx/platform#2359]
I've laid out my thoughts more fully in a post on [NGRX Offline patterns
](https://github.com/ngrx/platform/issues/994#issuecomment-583763989) but I propose the following new actions to support offline capabilities:

UPDATE_CACHE_SERVER_VALUE - overwrites the `originalValue` in `changeState` creating it if necessary and if the entity in the collection differs update the `changeType` to **Updated**. Leave the collection entity as is. **Use**: if we receive server data when we've already edited data i.e. we populate store from indexeddb and than receive server state.

UPDATE_COLLECTION_ENTITY - overwrites the collection entity leaving `changeState` except if the `changeType` is **Unchanged** and the collection entity and `originalValue` differ then we change it to **Updated**. **Use**: we've already received server data and we're merging from indexeddb

CLEAR_DELETED_ENTITY - clears the entity from the `collection` and `changeState` fully. **Use**: Server has deleted entity but we hold ngrx cache data.

SCHEDULE_FOR_DELETE - update `changeType` to **Deleted** and clear entity from the collection

UPSERT_OFFLINE - upserts indexeddb with a key, value pair of key `selectID(entity)` and value `{ entity, changeType, originalValue }`

DELETE_OFFLINE - clear key, value pair in indexeddb if exists.

## Other information:

I believe supporting offline to be the natural direction of @ngrx/data which could be implemented with the above. Have I missed a way to do the above with the existing api?

 
## If accepted, I would be willing to submit a PR for this feature

[ x ] Yes (Assistance is provided if you need help submitting a pull request)
[ ] No

**Comment by AdditionAddict** [comment]
I believe `MergeChanges` tracking will cover this requirement (after bugfixes).

I'll be exploring ways of implementing offline capabilities using current framework

**Comment by wardbell** [comment]
@AdditionAddict  We had offline scenarios in mind when we designed NgRx Data.  

I must begin with cautionary observations. I've been at this for two decades and I can assure you that data synchronization is really tricky.  I've never seen a solution that (a) was simple or (b) didn't stumble spectacularly at least some of the time. 

I do not believe I have the genius to solve a problem that has bewildered others. I would be leery of an attempt to complicate NgRx Data with a synchronization layer.

With that (discouraging) introduction, I hasten to add that I think NgRx Data already has all the primitive you need to satisfy many offline scenarios, including many you mentioned and the ones that I've encountered. Specifically

* All entities are in a single entity cache so you can easily access all or a subset of the entire data model in one place in a single select.
* You can easily serialize it to browser local storage.
* You have all the type and key information you'll need to store in IndexDB if that is your preference
* The embedded change tracking  allows you to separately select and serialize the unsaved changes and replay the later against either the same or a different cache (as long at the target caches have the same metadata).
* The change tracking merge semantics allow you to apply changes in a variety of ways, on a whole entity basis. The actions for this purpose are already in NgRx.
* If you decide that you'd like to merge parts of an entity (e.g., only merge the changed LastName but not the Social Security Number), you'll have to program that yourself. But you have the actions you'll need to apply changes after you've decide which parts of each entity should "win".

Hope that inspires you.

**Comment by AdditionAddict** [comment]
@wardbell Thank you for the above.

...is really tricky. I'm getting that :)

I agree everthing is there (I overlooked the [EntityDispatcher](https://ngrx.io/api/data/EntityDispatcher) API esp. use of mergeStrategy - your 2nd to last point) and offline could be built on top off this but not inside @ngrx/data which is why I closed the issue.

Big fixes - see PR #2370. I've not had a in depth look at the test suite but I'd be uncomfortable using mergeStrategy on methods that don't have documentation or testing but API suggests ability to use. I plan to put some testing in place. Additional #2307. BTW I'm not dissing the library - it's awesome and has helped me become a better programmer by being able to understand parts of the source code and has reduced errors on my non-offline project.

As an aside, I'm curious is the idea of keeping separate entityUnchangedCache, entityAddedCache, entityUpdatedCache and entityDeletedCache and making heavier use of @ngrx/entity came up as an under the hood refactor?
